src/goto-instrument/replace_function_bodies.cpp

```diff
@@ -0,0 +1,367 @@
+    auto assume_instruction = function.body.add_instruction();
+    assume_instruction->make_assumption(false_exprt());
+    assume_instruction->code = code_assumet(false_exprt());
+    auto end_instruction = function.body.add_instruction();
+    end_instruction->make_end_function();
+  }
+};
+
+class empty_replace_function_bodiest : public replace_function_bodiest
+{
+protected:
+  void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) override
+  {
+  }
+};
+
+class havoc_replace_function_bodiest : public replace_function_bodiest
+{
+public:
+  havoc_replace_function_bodiest(
+    std::vector<irep_idt> globals_to_havoc,
+    std::regex parameters_to_havoc)
+    : globals_to_havoc(std::move(globals_to_havoc)),
+      parameters_to_havoc(parameters_to_havoc)
+  {
+  }
+
+protected:
+  void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) override
+  {
+    namespacet ns(symbol_table);
+    for(auto const &parameter : function.type.parameters())
+    {
+      if(
+        parameter.type().id() == ID_pointer &&
+        !parameter.type().subtype().get_bool(ID_C_constant) &&
+        std::regex_match(
+          id2string(parameter.get_base_name()), parameters_to_havoc))
+      {
+        auto assign_instruction = function.body.add_instruction();
+        assign_instruction->make_assignment();
+        assign_instruction->code = code_assignt(
+          dereference_exprt(
+            symbol_exprt(parameter.get_identifier(), parameter.type())),
+          side_effect_expr_nondett(parameter.type().subtype()));
+      }
+    }
+
+    for(auto const &global_id : globals_to_havoc)
+    {
+      auto assign_instruction = function.body.add_instruction();
+      assign_instruction->make_assignment();
+      auto global_sym = ns.lookup(global_id);
+      assign_instruction->code = code_assignt(
+        symbol_exprt(global_sym.name),
+        side_effect_expr_nondett(global_sym.type));
+    }
+    if(function.type.return_type() != void_typet())
+    {
+      auto return_instruction = function.body.add_instruction();
+      return_instruction->make_return();
+      return_instruction->code =
+        code_returnt(side_effect_expr_nondett(function.type.return_type()));
+    }
+    auto end_function_instruction = function.body.add_instruction();
+    end_function_instruction->make_end_function();
+  }
+
+private:
+  const std::vector<irep_idt> globals_to_havoc;
+  std::regex parameters_to_havoc;
+};
+
+static std::vector<std::string>
+tokenize(const std::string &base, char separator)
+{
+  std::vector<std::string> tokens;
+  tokens.push_back("");
+  for(char c : base)
+  {
+    if(c == separator)
+    {
+      tokens.push_back("");
+    }
+    else
+    {
+      tokens.back().push_back(c);
+    }
+  }
+  return tokens;
+}
+
+static void write_to_stream(std::ostream &out)
+{
```

Comment:
huh?

Reply:
This is just to have a convenient method to assemble a string... Didn't end up actually needed it here so I'm removing it, but I figure something along those lines might be useful for string-utils?

Reply:
Maybe? I'll reserve judgement until I see a PR with it being used :stuck_out_tongue: